pkg/util/inplaceupdate/inplace_update_vertical.go

```diff
@@ -0,0 +1,252 @@
+
+func (v *VerticalUpdate) ValidateResourcePatch(op *jsonpatch.Operation, oldTemp *v1.PodTemplateSpec, updateSpec *UpdateSpec) error {
+	// for example: /spec/containers/0/resources/limits/cpu
+	words := strings.Split(op.Path, "/")
+	if len(words) != 7 {
+		return fmt.Errorf("invalid resource path: %s", op.Path)
+	}
+	idx, err := strconv.Atoi(words[3])
+	if err != nil || len(oldTemp.Spec.Containers) <= idx {
+		return fmt.Errorf("invalid container index: %s", op.Path)
+	}
+	quantity, err := resource.ParseQuantity(op.Value.(string))
+	if err != nil {
+		return fmt.Errorf("parse quantity error: %v", err)
+	}
+
+	if !v.CanResourcesResizeInPlace(words[6]) {
+		return fmt.Errorf("disallowed inplace update resouece: %s", words[6])
+	}
+
+	if _, ok := updateSpec.ContainerResources[oldTemp.Spec.Containers[idx].Name]; !ok {
+		updateSpec.ContainerResources[oldTemp.Spec.Containers[idx].Name] = v1.ResourceRequirements{
+			Limits:   make(v1.ResourceList),
+			Requests: make(v1.ResourceList),
+		}
+	}
+	switch words[5] {
+	case "limits":
+		updateSpec.ContainerResources[oldTemp.Spec.Containers[idx].Name].Limits[v1.ResourceName(words[6])] = quantity
+	case "requests":
+		updateSpec.ContainerResources[oldTemp.Spec.Containers[idx].Name].Requests[v1.ResourceName(words[6])] = quantity
+	}
+	return nil
+}
+
+// Get the resource status from the container and synchronize it to state
+func (v *VerticalUpdate) SyncContainerResource(container *v1.ContainerStatus, state *appspub.InPlaceUpdateState) {
+	if container == nil {
+		return
+	}
+
+	if state.LastContainerStatuses == nil {
+		state.LastContainerStatuses = make(map[string]appspub.InPlaceUpdateContainerStatus)
+	}
+	c := state.LastContainerStatuses[container.Name]
+	if state.LastContainerStatuses[container.Name].Resources.Limits == nil {
+		c.Resources.Limits = make(map[v1.ResourceName]resource.Quantity)
+	}
+	if state.LastContainerStatuses[container.Name].Resources.Requests == nil {
+		c.Resources.Requests = make(map[v1.ResourceName]resource.Quantity)
+	}
+
+	if container.Resources != nil {
+		for key, quantity := range container.Resources.Limits {
+			if !v.CanResourcesResizeInPlace(string(key)) {
+				continue
+			}
+			c.Resources.Limits[key] = quantity
+		}
+		for key, quantity := range container.Resources.Requests {
+			if !v.CanResourcesResizeInPlace(string(key)) {
+				continue
+			}
+			c.Resources.Requests[key] = quantity
+		}
+	}
+	// Store container infos whether c is empty or not.
+	// It will be used in IsContainerUpdateCompleted to check container resources is updated.
+	// case: pending pod can also be resize resource
+	state.LastContainerStatuses[container.Name] = c
+}
+
+// UpdateResource implements vertical updates by directly modifying the container's resources,
+// conforming to the k8s community standard
+func (v *VerticalUpdate) UpdateContainerResource(container *v1.Container, newResource *v1.ResourceRequirements) {
+	if container == nil || newResource == nil {
```

Comment:
we can not return directly since it is possible to set or unset container resource for a pod of burstable qos class

Reply:
We should recreate pod when try to change qos class.